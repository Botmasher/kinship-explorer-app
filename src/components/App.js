import React, { Component } from 'react';
import { Route, Switch } from 'react-router-dom';
import GameContainer from './GameContainer';
import PageNotFound from './PageNotFound';
import { store } from '../store';
import { gameExists, isGameLoaded, onGameLoaded } from '../utils';

class App extends Component {
	constructor(props) {
		super(props);
		this.state = {
			isLoaded: false,
			unloadedClicks: 0
		};
	}

	handleLoadingClick = () => {
		// check for routing during game load - used to update div keys
		isGameLoaded()
			? this.setState({isLoaded: true})
			: this.setState(prevState => ({unloadedClicks: prevState.unloadedClicks+1}))
		;
	};

	componentDidMount() {
		// check that game has fully loaded (used for e.g. mounting menu description)
		!this.state.isLoaded && onGameLoaded(() => this.setState({isLoaded: true}));
	}

	render() {
		const { isLoaded, unloadedClicks } = this.state;
		const { systems } = store;
		return (
			<div className="App">
				<h1 className="app-title">
					<span className="letter-decoration">K</span>{'inship '}
					<span className="letter-decoration">T</span>{'erm '}
					<span className="letter-decoration">E</span>{'xplorer'}
				</h1>
				{gameExists()
					?	<Switch>
							<Route exact path="/" render={() => (
								<GameContainer
									handleLoadingClick={this.handleLoadingClick}
									systemId={'global'}
									isGameLoaded={isLoaded}
									unloadedClicks={unloadedClicks}
								/>
							)} />
							<Route exact path="/:system?" render={({ match }) => (
								match && match.params && Object.keys(systems).includes(match.params.system)
								?	<GameContainer
										handleLoadingClick={this.handleLoadingClick}
										systemId={match && match.params && match.params.system ? match.params.system : null}
										isGameLoaded={isLoaded}
										unloadedClicks={unloadedClicks}
									/>
								: <PageNotFound match={match} />
							)} />
							<Route component={PageNotFound} />
						</Switch>
					: <div>
							<p>
								My game player isn't loading on your browser or device. Do you happen to have one that{' '}
								<a href="https://docs.unity3d.com/Manual/webgl-browsercompatibility.html">supports Unity running on WebGL</a>?
							</p>
						</div>
				}

			</div>
		);
	}
}

export default App;
