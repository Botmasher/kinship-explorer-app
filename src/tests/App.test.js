import React from 'react';
import ReactDOM from 'react-dom';
import App from '../components/App';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
	const div = document.createElement('div');
	ReactDOM.render(<App />, div);
	ReactDOM.unmountComponentAtNode(div);
});

it('renders a snapshot', () => {
	const tree = renderer.create(<App />).toJSON();
	expect(tree).toMatchSnapshot();
});

it('handles loading clicks', () => {
	const component = renderer.create(
		<App />
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();

	tree.props.handleLoadingClick();
	expect(tree).toMatchSnapshot();
});
