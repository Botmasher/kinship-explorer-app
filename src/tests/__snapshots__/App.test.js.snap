// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle loading clicks 1`] = `
<div
  className="App"
>
  <h1
    className="app-title"
  >
    <span
      className="letter-decoration"
    >
      K
    </span>
    inship 
    <span
      className="letter-decoration"
    >
      T
    </span>
    erm 
    <span
      className="letter-decoration"
    >
      E
    </span>
    xplorer
  </h1>
  <div>
    <p>
      My game player isn't loading on your browser or device. Do you happen to have one that
       
      <a
        href="https://docs.unity3d.com/Manual/webgl-browsercompatibility.html"
      >
        supports Unity running on WebGL
      </a>
      ?
    </p>
  </div>
</div>
`;

exports[`handles loading clicks 1`] = `
<div
  className="App"
>
  <h1
    className="app-title"
  >
    <span
      className="letter-decoration"
    >
      K
    </span>
    inship 
    <span
      className="letter-decoration"
    >
      T
    </span>
    erm 
    <span
      className="letter-decoration"
    >
      E
    </span>
    xplorer
  </h1>
  <div>
    <p>
      My game player isn't loading on your browser or device. Do you happen to have one that
       
      <a
        href="https://docs.unity3d.com/Manual/webgl-browsercompatibility.html"
      >
        supports Unity running on WebGL
      </a>
      ?
    </p>
  </div>
</div>
`;

exports[`renders a snapshot 1`] = `
<div
  className="App"
>
  <h1
    className="app-title"
  >
    <span
      className="letter-decoration"
    >
      K
    </span>
    inship 
    <span
      className="letter-decoration"
    >
      T
    </span>
    erm 
    <span
      className="letter-decoration"
    >
      E
    </span>
    xplorer
  </h1>
  <div>
    <p>
      My game player isn't loading on your browser or device. Do you happen to have one that
       
      <a
        href="https://docs.unity3d.com/Manual/webgl-browsercompatibility.html"
      >
        supports Unity running on WebGL
      </a>
      ?
    </p>
  </div>
</div>
`;
